Kernel/Storage/ATA/ATAPort.cpp

```diff
@@ -0,0 +1,459 @@
+    }
+    return {};
+}
+
+RefPtr<StorageDevice> ATAPort::connected_device(size_t device_index) const
+{
+    MutexLocker locker(m_lock);
+    if (m_ata_devices.size() > device_index)
+        return m_ata_devices[device_index];
+    return {};
+}
+
+ErrorOr<void> ATAPort::start_request(const ATADevice& associated_device, AsyncBlockDeviceRequest& request)
+{
+    MutexLocker locker(m_lock);
+    VERIFY(m_current_request.is_null());
+    VERIFY(pio_capable() || dma_capable());
+
+    dbgln_if(ATA_DEBUG, "ATAPort::start_request");
+
+    m_current_request = request;
+    m_current_request_block_index = 0;
+    m_current_request_flushing_cache = false;
+
+    if (dma_capable()) {
+        TRY(prepare_and_initiate_dma_transaction(associated_device));
+        return {};
+    }
+    TRY(prepare_and_initiate_pio_transaction(associated_device));
+    return {};
+}
+
+void ATAPort::complete_pio_transaction(AsyncDeviceRequest::RequestResult result)
+{
+    VERIFY(m_current_request);
+
+    // Now schedule reading back the buffer as soon as we leave the irq handler.
+    // This is important so that we can safely write the buffer back,
+    // which could cause page faults. Note that this may be called immediately
+    // before Processor::deferred_call_queue returns!
+    g_io_work->queue([this, result]() {
+        dbgln_if(ATA_DEBUG, "ATAPort::complete_pio_transaction result: {}", (int)result);
+        MutexLocker locker(m_lock);
+        VERIFY(m_current_request);
+        auto current_request = m_current_request;
+        m_current_request.clear();
+        current_request->complete(result);
+    });
+}
+
+void ATAPort::complete_dma_transaction(AsyncDeviceRequest::RequestResult result)
+{
+    // NOTE: this may be called from the interrupt handler!
+    VERIFY(m_current_request);
+    VERIFY(m_lock.is_locked());
+
+    // Now schedule reading back the buffer as soon as we leave the irq handler.
+    // This is important so that we can safely write the buffer back,
+    // which could cause page faults. Note that this may be called immediately
+    // before Processor::deferred_call_queue returns!
+    g_io_work->queue([this, result]() {
+        dbgln_if(ATA_DEBUG, "ATAPort::complete_dma_transaction result: {}", (int)result);
+        MutexLocker locker(m_lock);
+        VERIFY(m_current_request);
+        auto current_request = m_current_request;
+        m_current_request.clear();
+
+        if (result == AsyncDeviceRequest::Success) {
+            {
+                auto result = force_busmastering_status_clean();
+                // FIXME: Propagate errors properly
+                VERIFY(!result.is_error());
+            }
+
+            if (current_request->request_type() == AsyncBlockDeviceRequest::Read) {
+                if (auto result = current_request->write_to_buffer(current_request->buffer(), m_dma_buffer_region->vaddr().as_ptr(), 512 * current_request->block_count()); result.is_error()) {
+                    //m_lock.unlock();
```

Comment:
I think this should not be a comment anymore? I did have a deadlock situation in NVMe driver without unlocking the lock before completing the request. 
Same applies to complete request at the end